recursion/core/src/poseidon2/external.rs

```diff
@@ -50,56 +71,90 @@ impl<F: PrimeField32> MachineAir<F> for Poseidon2Chip {
     }
 
     #[instrument(name = "generate poseidon2 trace", level = "debug", skip_all)]
+    #[allow(clippy::needless_range_loop)]
     fn generate_trace(
         &self,
-        _: &ExecutionRecord<F>,
-        _: &mut ExecutionRecord<F>,
+        input: &ExecutionRecord<F>,
+        _output: &mut ExecutionRecord<F>,
     ) -> RowMajorMatrix<F> {
-        let mut input = [F::one(); WIDTH];
-        let rows = (0..128)
-            .map(|i| {
+        let mut rows = Vec::new();
+        // let mut record = ExecutionRecord::default();
+
+        for i in 0..input.poseidon2_events.len() {
+            let event = input.poseidon2_events[i].clone();
+
+            // init state for the first round.
+            // this will also be used to store the pre-state for the next round.
+            let mut pre_state = event.initial_state;
+
+            let memory = event.initial_state_records.clone();
+
+            for i in 0..NUM_TOTAL_ROUNDS {
                 let mut row = [F::zero(); NUM_POSEIDON2_COLS];
                 let cols: &mut Poseidon2Cols<F> = row.as_mut_slice().borrow_mut();
-                cols.input = input;
 
-                let r = i % 31;
-                let rounds_f = 8;
-                let rounds_p = 22;
-                let rounds = rounds_f + rounds_p;
-                let rounds_f_beginning = rounds_f / 2;
-                let p_end = rounds_f_beginning + rounds_p;
+                // increment the clock.
+                cols.clk = event.clk + F::from_canonical_usize(i);
+
+                cols.state_ptr = event.state_ptr;
+
+                // If this is the first round, we need to initialize the memory records.
+                if i == 0 {
+                    for j in 0..POSEIDON2_WIDTH {
+                        cols.state[j].populate(&event.initial_state_records[j]);
+                    }
+
+                    cols.do_memory_check = F::one();
+                    cols.first_round = F::one();
+
+                    // this row is only for memory readfing.
+                    continue;
+                } else {
+                    for j in 0..POSEIDON2_WIDTH {
+                        cols.state[j].populate(&MemoryRecord {
+                            addr: cols.state[j].addr,
+                            value: memory[j].value,
```

Comment:
what's going on here? is this just a dummy memory record?

Reply:
to make the code cleaner, i would suggest doing an actual memory read/write on every row that way you will not need to handle complex flags.

like in row 1, read from ptr write to ptr
in row 2, read from ptr write to ptr
etc.

the cost will be the same